Code Movement to Prod

LBWright opened the pull request:
# Changelog

### REST API URIs for:
* Users
* Houses

### Authentication methods:
* JWT
* User Id inside of JWT Payload

### API Feature Implementation:
* User is verified and added to request object
* We have an error handler that handles errors based on their statuses

### Client Side:
* Application architecture complete and modules installed

LBWright committed:
Merge pull request #2 from Lambda-School-Labs/feature/setup/server

Feature/setup/server

BryceEvans committed:
react app created

mglittleton committed:
added dependencies for react and redux

johnoro committed:
Merge pull request #3 from Lambda-School-Labs/front-end/create-react-app

Creates React App

BryceEvans committed:
React app created and Redux dependencies added.

BryceEvans committed:
React app created and Redux dependencies added.

LBWright committed:
Merge pull request #8 from Lambda-School-Labs/front-end/create-react-app

Front end/create react app without rental folder

kybak committed:
Merge branch 'feature/setup/server' of https://github.com/Lambda-School-Labs/labspt2-rental into feature/setup/server

kybak committed:
clone and added .idea to .gitignore

juliejonak committed:
Added wireframes:
- Dashboard: dashboard component, owner and employee views
- Nav Bar: notifications and help
- Properties: owner and employee views, add property, view property

juliejonak committed:
Replaced Add Property Page 1 with updated version

jessharrison83 committed:
Merge pull request #9 from Lambda-School-Labs/mockups

Mockups: Dashboard, NavBar and Properties

juliejonak committed:
Settings wireframes added

jessharrison83 committed:
Merge pull request #10 from Lambda-School-Labs/mockups

Mockups: Settings

LBWright committed:
Merge branch 'feature/setup/server' of https://github.com/Lambda-School-Labs/labspt2-rental into development

mglittleton committed:
added main file structure for some pages

Properties, Home, and Settings pages as well as two types of nav bars

mglittleton committed:
started redux formatting

mglittleton committed:
removed node modules

mglittleton committed:
commented out Redux temporarily

johnoro committed:
Merge pull request #11 from Lambda-School-Labs/front-end/folder-structure

added main file structure for some pages

jessharrison83 committed:
Top Nav initial commit

component created for top nav bar, starter code in app.js removed, top nav component added to app.js to make sure it renders. text renders correctly.

jessharrison83 committed:
Top Nav Skeleton

basic component built, no functionality yet. very basic styling to place elements in the correct spot done. still needs a logo image. icons folder created.

mglittleton committed:
added basic redux structure to index.js and to Redux/store.js

incorporates the blank reducers.js file in the PropertyList component directory

johnoro committed:
Merge pull request #12 from Lambda-School-Labs/top-nav

Top nav skeleton created

johnoro committed:
Merge pull request #13 from Lambda-School-Labs/redux-setup

added basic redux structure to index.js and to Redux/store.js

LBWright committed:
Merge branch 'development' of https://github.com/Lambda-School-Labs/labspt2-rental into development

LBWright committed:
added house resources

juliejonak committed:
Added wireframes:
- Employee dashboard views for owner and employees with permissions
- Reservation dashboard views for varied permissions
- Tasks dashboard views for employee and owner permissions
- Updated dashboard home views from employees and owner

jessharrison83 committed:
Merge pull request #15 from Lambda-School-Labs/mockups

Added new wireframes:

jessharrison83 committed:
Merge pull request #6 from Lambda-School-Labs/feature/setup/server

initial server commit

jessharrison83 committed:
Merge branch 'development' into features/resource/house

juliejonak committed:
Merge pull request #14 from Lambda-School-Labs/features/resource/house

Features/resource/house

LBWright committed:
added start script to root project directory

LBWright committed:
added start script to root project directory

jessharrison83 committed:
Reservations API

Controller, model, and router file created for reservations resource.

johnoro commented on client/src/Components/shared/TopNavBar/container.js:
```diff
@@ -0,0 +1,27 @@
+            src={require("./icons/notification-bell.svg")}
+            alt="Notifications"
+          />
+          <img src={require("./icons/help-button.svg")} alt="Help" />
```
I have to wonder how well these requires in the code will scale. It seems like a lot of effort to type it out, and what if you want to conveniently change your icons directory? Let's see an example function for this common `require`:
```javascript
class TopNav extends React.Component {
  requireSvg = name => require(`./icons/${name}.svg`);
  // in your render's return...
    <img src={this.requireSvg("menu")} alt="descriptive text" />
}
```

LBWright replied:
Yeah. This will probably need to be removed. This isn't what the container is for.

LBWright replied:
@mglittleton

LBWright replied:
@BryceEvans

jessharrison83 replied:
This is mine. I'm not sure I understand what you mean by "this isn't what the container is for."

LBWright replied:
@jessharrison83 Did our convo resolve this question? I was just seeing it.

jessharrison83 replied:
Yes it did. I’ll fix this tonight.

johnoro commented on server/src/config/dev.js:
```diff
@@ -0,0 +1,6 @@
+export const config = {
+    secrets: {
+      jwt: 'mentor and stuff'
```
I assume these 'secrets' are just development placeholders and are meant to be pushed up?

LBWright replied:
yeah. These are local secrets.

johnoro commented on server/src/server.js:
```diff
@@ -0,0 +1,34 @@
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(morgan('dev'));
+app.use(express.json());
+app.use(bodyParser.urlencoded({ extended: false }));
```
It looks like there's some repeated code here; there isn't a purpose for this, is there?

LBWright replied:
Nope. We'll need to remove it. That looks like it got through a code review

johnoro commented on server/src/utils/errorHandler.js:
```diff
@@ -0,0 +1,23 @@
+      res.status(500).json({ error: err.message, success: false })
+      break
+    default:
+      res.status(500).json({ error: err.message, success: false })
```
It looks like there's a lot of repeated code here; the only thing that's changing at the moment is the statusCode passed to `status`. It's understandable if this is going to be expanded on and customized later, but, if not, then I'd try to get rid of the unnecessary repetition!

LBWright replied:
Actually, this avoids the repetition. The case statement lets us throw custom errors on the frontend, pass the status code down and then send it back up. All of the errors are handled in this switch statement.

johnoro replied:
I'm confused with the code itself if it isn't going to be expanded further, but I might just be too tired to realize how this isn't repetitive. It _looks_ like you could do the following in place of the switch statement:
```javascript
res.status(err.statusCode).json({ error: err.message, success: false })
```

LBWright replied:
Oh yeah. Okay. So there are two reasons for this. The first is just in case we want to do additional logic based on the error type, but the second is that sometimes Mongoose sends back non http standard error codes (which won't have a code) when there are validation errors and we want to send those back as 500 errors and handle them all the same.

I've thought about doing this
```js
if (!err.statusCode) err.statusCode = 500
res.status(err.statusCode).json({ error: err.message })
```

johnoro replied:
The non http standard is good to know, and the additional logic is definitely what would validate the switch statement if nothing else!

johnoro reviewed (approved):
Looks good to me.

mglittleton committed:
added invoice items API

API includes: name, description, price, owner, defaultItem (Boolean), lastUsed

LBWright reviewed (commented):
Okay. Don't merge this yet then. We'll need to address some of these issues.

johnoro commented:
This is looking awesome regardless, let me know when you want it merged!

BryceEvans committed:
task API and Tests Written

mglittleton committed:
adds tests to the invoice_items API

BryceEvans committed:
comments deleted and task tests passing. It makes a difference to install mongoDB and get the server running.

BryceEvans committed:
spacing adjusted on task.model.spec

BryceEvans committed:
toml file added in prepeartion for deployment.

juliejonak committed:
Merge pull request #21 from Lambda-School-Labs/dashboard

toml file added in prepeartion for deployment.

BryceEvans committed:
toml file changed.

juliejonak committed:
Merge pull request #22 from Lambda-School-Labs/dashboard

toml file changed.

BryceEvans committed:
toml changes again for deployment.

BryceEvans committed:
more toml changes.

juliejonak committed:
Merge pull request #23 from Lambda-School-Labs/dashboard

Netlify deploy is live!

LBWright committed:
rearranged redux and components

LBWright committed:
Merge branch 'development' into feature/client/registration

juliejonak committed:
Boilerplate file structure

juliejonak committed:
Filling in API:
- controllers extended from BaseController.js
- router.js: routes created
- model.js: shell code added

juliejonak committed:
Filled in billingSchema

juliejonak committed:
Added testing

LBWright committed:
Merge pull request #20 from Lambda-School-Labs/tasks-api

task API and Tests Written (& passing)!

LBWright committed:
Merge pull request #19 from Lambda-School-Labs/server/invoice-items-api

added invoice items API

juliejonak committed:
Adjust name to billingPlan

LBWright committed:
Merge pull request #17 from Lambda-School-Labs/reservation-api

Reservations API

LBWright committed:
Merge pull request #25 from Lambda-School-Labs/billingAPI

Billing API

LBWright committed:
Merge pull request #24 from Lambda-School-Labs/feature/client/registration

rearranged redux and components

BryceEvans committed:
edited toml file to npm run build.

LBWright committed:
Merge pull request #26 from Lambda-School-Labs/edit-toml

edited toml file to npm run build.

jessharrison83 committed:
userList component

component created, route added, reducer added to root reducer. functionality still needs to be completed.

BryceEvans committed:
property api added but not all tests passing just yet.

jessharrison83 committed:
actions and reducer completed

juliejonak committed:
Added LoginPage

mglittleton committed:
adds discount API

discount API includes name, description, discountPercentage, owner, lastUsed

LBWright committed:
created registration component

BryceEvans committed:
Property Endpoints Written & Tests are Passing

mglittleton committed:
fixes several API errors in discounts and invoice_items

These were mostly due to typo errors I made

LBWright committed:
Merge pull request #29 from Lambda-School-Labs/feature/registration

created registration component

LBWright committed:
Merge pull request #28 from Lambda-School-Labs/server/discount-api

adds discount API

BryceEvans committed:
added routes to router.js (I think) and property & task tests all still passing.

jessharrison83 committed:
Update actions.js

BryceEvans committed:
Okay, now I added routes to router.js and task and property tests still passing.

jessharrison83 committed:
Update reducer.js

jessharrison83 committed:
Update actions.js

LBWright committed:
Merge pull request #30 from Lambda-School-Labs/server/property-resource-api

Server/property resource api

LBWright committed:
Merge pull request #27 from Lambda-School-Labs/user-list

User list

juliejonak committed:
Cleaned up Login Page component

juliejonak committed:
Adjusted LoginPage actions and reducers

LBWright committed:
added RegistrationPage snapshot and test

kybak committed:
set up theme

mglittleton committed:
Merge pull request #16 from Lambda-School-Labs/feature/startscript

Feature/startscript

mglittleton committed:
Merge pull request #32 from Lambda-School-Labs/feature/login

Login Page Feature

mglittleton committed:
Merge pull request #34 from Lambda-School-Labs/semantic-theme

Initialize Semantic Theme

kybak committed:
flex boxes

kybak committed:
local dependency custom-components

kybak committed:
remove node_modules

jessharrison83 committed:
Merge pull request #33 from Lambda-School-Labs/feature/registration

added RegistrationPage snapshot and test

kybak committed:
Merge branch 'development' into registration-page

jessharrison83 committed:
Merge pull request #39 from Lambda-School-Labs/registration-page

Registration Page
